Targeting Optimizations (#1364)
Phobos-developers/Phobos: docs/Fixed-or-Improved-Logics.md

```

```{warning}

> Auto-deploy feature requires `Convert.Deploy` from [Ares 1�7 Type Conversion](https://ares-developers.github.io/Ares-docs/new/typeconversion.html) to change type. Unit without it will constantly use deploy command on self until ammo is changed.

Reviewer: ```suggestion
Auto-deploy feature requires `Convert.Deploy` from [Ares' Type Conversion](https://ares-developers.github.io/Ares-docs/new/typeconversion.html) to change type. Unit without it will constantly use deploy command on self until ammo is changed.
```

IsSingleColor=false  ; boolean
```

> ### Target scanning delay optimization
> In Vanilla, the game used 'NormalTargetingDelay' and 'GuardAreaTargetingDelay' to globally control the target searching intervals. Increasing these values would make units stupid, while decreasing them would cause the game lagging.
> Now, you can define them per techno, also allowing different values for AI and players. The default values are the same as those originally defined by the vanilla flags.

Reviewer: ```suggestion
### Target scanning delay optimization

- In vanilla, the game used 'NormalTargetingDelay' and 'GuardAreaTargetingDelay' to globally control the target searching intervals. Increasing these values would make units stupid, while decreasing them would cause game lag.
- Now, you can define them per techno, also allowing different values for AI and players. The default values are the same as those originally defined by the vanilla flags.
```

>  In `rulesmd.ini`:
> ```ini
> [General]
> AINormalTargetingDelay=              ;integer, game frames
> PlayerNormalTargetingDelay=          ;integer, game frames
> AIGuardAreaTargetingDelay=           ;integer, game frames
> PlayerGuardAreaTargetingDelay=       ;integer, game frames
> 
> [SOMETECHNO]
> AINormalTargetingDelay=              ;integer, game frames
> PlayerNormalTargetingDelay=          ;integer, game frames
> AIGuardAreaTargetingDelay=           ;integer, game frames
> PlayerGuardAreaTargetingDelay=       ;integer, game frames
> ```

Reviewer: ```suggestion
In `rulesmd.ini`:
```ini
[General]
AINormalTargetingDelay=              ; integer, game frames
PlayerNormalTargetingDelay=          ; integer, game frames
AIGuardAreaTargetingDelay=           ; integer, game frames
PlayerGuardAreaTargetingDelay=       ; integer, game frames

[SOMETECHNO]                         ; TechnoType
AINormalTargetingDelay=              ; integer, game frames
PlayerNormalTargetingDelay=          ; integer, game frames
AIGuardAreaTargetingDelay=           ; integer, game frames
PlayerGuardAreaTargetingDelay=       ; integer, game frames
```
```

FireOnce.ResetSequence=true  ; boolean
```

### Force techno targeting in distributed frames to improve performance

> - When you create many techno in a same frame (f.ex. starting the game with a campaign map that initially has a large number of technos), they will always scan for target in a synchronous period, causing the game lagging. Increasing targeting delay will not make things better, as their targeting are still synchronized.

Reviewer: ```suggestion
- When you create many technos in a same frame (i.e. starting the game with a campaign map that initially has a large number of technos), they will always scan for targets in a synchronous period, causing game lag. Increasing targeting delay alone will not make things better, as their targeting is still synchronized.
```

> - It is now possible to force them targeting seperatly. When a techno spawn, they generate a random number in \[0,15\].
> - If `DistributeTargetingFrame=true` is set, only when the current frame number is congruent with the technos own number under modulo 16, will it do targeting.
> - You can use `DistributeTargetingFrame.AIOnly` to make it only work for AI (Players are not likely to have so many technos.).
> - For technos with low ROF (f.ex. Tanya), this will significantly reduce their efficiency in killing enemies. Turn the same flag off on techno to prevent it.

Reviewer: ```suggestion
- It is now possible to force them to seek targets separately. When a techno spawns, it will generate a random number in \[0,15\]. If `DistributeTargetingFrame=true` is set, only when the current frame number is congruent with the technos own number under modulo 16, will it do targeting.
  - You can use `DistributeTargetingFrame.AIOnly` to make it only work for AI (Players are not likely to have so many technos.)
  - For technos with low ROF (i.e. Tanya), this will significantly reduce their efficiency in killing enemies.
```

In `rulesmd.ini`
```ini
[General]
DistributeTargetingFrame=false         ; boolean
DistributeTargetingFrame.AIOnly=true   ; boolean

> [SomeTechno]
> DistributeTargetingFrame=              ; boolean

Reviewer: ```suggestion
[SOMETECHNO]                           ; TechnoType
DistributeTargetingFrame=              ; boolean
```
